app/models/frontend/nics/NICsAnswers.scala

```diff
@@ -17,15 +17,49 @@
 package models.frontend.nics
 
 import cats.implicits.catsSyntaxOptionId
+import models.common.BusinessId
 import models.connector.api_1639.SuccessResponseAPI1639
+import models.connector.api_1803
+import models.connector.api_1803.AnnualNonFinancialsType
 import models.database.nics.NICsStorageAnswers
+import models.frontend.nics.NICsClass4Answers.Class4ExemptionAnswers
 import play.api.libs.json.{Format, Json}
 
 case class NICsAnswers(class2Answers: Option[NICsClass2Answers], class4Answers: Option[NICsClass4Answers])
 
 object NICsAnswers {
   implicit val formats: Format[NICsAnswers] = Json.format[NICsAnswers]
 
+  def mkPriorData(maybeDisclosures: Option[SuccessResponseAPI1639],
+                  maybeAnnualSummaries: List[(api_1803.SuccessResponseSchema, BusinessId)],
+                  maybeDbAnswers: Option[NICsStorageAnswers]): Option[NICsAnswers] = {
+
+    val class2Data: Option[Boolean]              = maybeDisclosures.flatMap(_.class2Nics.flatMap(_.class2VoluntaryContributions))
+    val class2DatabaseData: Option[Boolean]      = maybeDbAnswers.flatMap(_.class2NICs)
+    val class4Data: List[Class4ExemptionAnswers] = mkClass4ExemptionData(maybeAnnualSummaries)
+    (class2Data, class2DatabaseData, class4Data) match {
+      case (_, _, class4Data) if class4Data.nonEmpty => mkPriorClass4Data(class4Data)
+      case (Some(true), _, _)                        => Some(NICsAnswers(class2Answers = NICsClass2Answers(true).some, None))
+      case (None, Some(false), _)                    => Some(NICsAnswers(class2Answers = NICsClass2Answers(false).some, None))
+      case _                                         => None
+    }
+  }
+
+  def mkClass4ExemptionData(multipleAnnualSummaries: List[(api_1803.SuccessResponseSchema, BusinessId)]): List[Class4ExemptionAnswers] = {
+    val exemptionList: List[Option[Class4ExemptionAnswers]] = multipleAnnualSummaries.map { case (answers, businessId) =>
+      val maybeClass4Exemption: Option[Boolean] = answers.annualNonFinancials.flatMap(_.exemptFromPayingClass4Nics)
+      maybeClass4Exemption.map { class4Exempt =>
+        val class4ExemptReason: Option[ExemptionReason] = {
+          val reason: Option[AnnualNonFinancialsType.Class4NicsExemptionReason.Value] =
+            answers.annualNonFinancials.flatMap(_.class4NicsExemptionReason)
+          reason.map(ExemptionReason.fromNonFinancialType)
+        }
+        Class4ExemptionAnswers(businessId, class4Exempt, class4ExemptReason)
+      }
+    }
+    exemptionList.flatten
+  }
```

Comment:
Could use `.flatMap` instead of `.map` and remove `exemptionList.flatten`:
```suggestion
  def mkClass4ExemptionData(multipleAnnualSummaries: List[(api_1803.SuccessResponseSchema, BusinessId)]): List[Class4ExemptionAnswers] = {
    multipleAnnualSummaries.flatMap { case (answers, businessId) =>
      val maybeClass4Exemption: Option[Boolean] = answers.annualNonFinancials.flatMap(_.exemptFromPayingClass4Nics)
      maybeClass4Exemption.map { class4Exempt =>
        val class4ExemptReason: Option[ExemptionReason] = {
          val reason: Option[AnnualNonFinancialsType.Class4NicsExemptionReason.Value] =
            answers.annualNonFinancials.flatMap(_.class4NicsExemptionReason)
          reason.map(ExemptionReason.fromNonFinancialType)
        }
        Class4ExemptionAnswers(businessId, class4Exempt, class4ExemptReason)
      }
    }
  }
```

Reply:
Yes @kuldeepwadhwa9 , would be good to make this change